graphistry/embed_utils.py

```diff
@@ -68,22 +73,31 @@ def embed(self, proto='TransE', d=128, gamma=12):
 
                 loss = em_mod(pos_triples, neg_triples)
                 loss.mean().backward()
-                print(loss.mean().item()) 
+                loss_acc += loss.mean().item()
+                c += 1
                 optim.step()
 
-        # out vectors
-        # --------------------------------------
-        # exception handling will be added later
+            epoch.set_description(f"loss: {loss_acc/c}")
+            epoch.refresh()
+
+        self.node_em = em_mod.node_em.weight.detach().numpy()
+        self.edge_em = em_mod.edge_em.weight.detach().numpy()
+        return self.node_em, self.edge_em
 
 
     def TransE(self, h, r, t):
-        gamma = 12
-        o = h + r - t
-        return gamma - (o/o.max()) # will be replaced by 2 norm later
+        return (h + r - t).norm(p=1, dim=1)
 
     def TransD(self, h, r, t):
         return
 
+    def DistMult(self, h, r, t):
```

Comment:
interest q: how to let users override their own  loss func, dist func, objective, etc

Reply:
Great q:  perhaps 

```from torch import nn
g.embed(model: ConcreteLossType = 'str' | nn.LogSigmoid )```

with ConcreteLossType checking that, if not string, it sends in a proper loss function (y_true, y_predicted)

Reply:
added custom distance function support, now you can use something like this
```python3
def custom_dist(h, r, t):
    return (h + r - t).norm(dim=1)

G.embed(proto=custom_dist)
```